internal/machine/update.go

```diff
@@ -18,6 +18,76 @@ func Update(ctx context.Context, m *api.Machine, input *api.LaunchMachineInput)
 		colorize    = io.ColorScheme()
 	)
 
+	if input != nil && input.Config != nil && input.Config.Guest != nil {
+		// Check that there's a valid number of CPUs
+		var validNumCpus []int
+
+		if input.Config.Guest.CPUKind == "shared" {
+			validNumCpus = append(validNumCpus, 1, 2, 4, 6, 8)
+
+		} else if input.Config.Guest.CPUKind == "performance" {
+			validNumCpus = append(validNumCpus, 1, 2, 4, 6, 8, 10, 12, 14, 16)
+
+		}
+
+		validCpuNum := false
+
+		for _, num := range validNumCpus {
+			if num == input.Config.Guest.CPUs {
+				validCpuNum = true
+				break
+
+			}
+		}
+
+		if !validCpuNum {
+			return fmt.Errorf("invalid config: invalid number of CPUs for %s guest. Valid numbers are %v\nView more information here: https://fly.io/docs/about/pricing/#machines", input.Config.Guest.CPUKind, validNumCpus)
+
+		}
+
+		if input.Config.Guest.CPUKind == "shared" && input.Config.Guest.MemoryMB%256 != 0 {
+			return fmt.Errorf("invalid config: invalid memory size; must be in 256 MiB increment\nView more information here: https://fly.io/docs/about/pricing/#machines")
+
+		} else if input.Config.Guest.CPUKind == "performance" && input.Config.Guest.MemoryMB%1024 != 0 {
+			return fmt.Errorf("invalid config: invalid memory size; must be in 1024 MiB increment\nView more information here: https://fly.io/docs/about/pricing/#machines")
+
+		}
+
+		var presetName string
+
+		if input.Config.Guest.CPUKind == "shared" {
+			presetName = fmt.Sprintf("shared-cpu-%dx", input.Config.Guest.CPUs)
+		} else if input.Config.Guest.CPUKind == "performance" {
+			presetName = fmt.Sprintf("performance-%dx", input.Config.Guest.CPUs)
+		}
+
+		// Check memory sizes
+		if machinePreset, ok := api.MachinePresets[presetName]; ok {
+			if machinePreset.MemoryMB > input.Config.Guest.MemoryMB {
+				return fmt.Errorf("invalid config: for machines with %d CPUs, the minimum amount of memory is %d MiB\nView more information here: https://fly.io/docs/about/pricing/#machines", machinePreset.CPUs, machinePreset.MemoryMB)
+
+			}
+
+			var maxMemory int
+
+			if input.Config.Guest.CPUKind == "shared" {
+				maxMemory = input.Config.Guest.CPUs * api.MAX_MEMORY_MB_PER_SHARED_CPU
+			} else if input.Config.Guest.CPUKind == "performance" {
+				maxMemory = input.Config.Guest.CPUs * api.MAX_MEMORY_MB_PER_CPU
+			}
+
+			if input.Config.Guest.MemoryMB > maxMemory {
+				return fmt.Errorf("invalid config: for machines with %d CPUs, the maximum amount of memory is %d MiB\nView more information here: https://fly.io/docs/about/pricing/#machines", machinePreset.CPUs, maxMemory)
+
+			}
+
+		} else {
+			// this shouldn't happen
```

Comment:
if getting here is a bug, reporting to sentry is a good way to ensure we see that the bug is happening.

Reply:
Actually now that you point this out, this shouldn't ever happen, because of previous checks. I'll make a new PR that takes that into account.